src/box/box.cc

```diff
@@ -412,6 +415,8 @@ box_do_set_orphan(bool orphan)
 {
 	is_orphan = orphan;
 	box_update_ro_summary();
+	if (!is_orphan)
+		BOX_ON_RECOVERY_STATE("synced");
```

Comment:
AFAIR the instance can become orphan again on `box.cfg` reconfig with new replication parameter. Is it true? If so, is it left out intentionally?

Reply:
> AFAIR the instance can become orphan again on box.cfg reconfig with new replication parameter

True.

> If so, is it left out intentionally?

What's left out? Once the instance syncs with new replication parameter, `'synced'` triggers will run once again.
Or do you mean there are no triggers on becoming an orphan?

Reply:
>  Btw, I just thought - why do we make these recovery states as triggers? Wouldn't a new built-in event work? Similar to "box.status".

The main difference between watchers and triggers is that triggers are run synchronously, and watcher callbacks are run asynchronously. So using watchers here would either mean that the events are delivered when box.cfg yields (which means we lose the exact moment when event happened), or we have to add more yields where the notification is sent.

So it seems triggers suit this particular use case more. For example, one might want to set up triggers on xlog recovery, without touching snapshot data.

> Yes. Maybe any of these states shouldn't be fired more than once?

I agree. Since the triggeres are called `on_recovery_state`, they should be run only during "recovery" or initial sync.
I'll add a run once guard for "synced" stage. All the other stages are already run once.

Reply:
Here. I've also added an enumeration of all states, so calling a trigger with some specific state is less error-prone now.

<details><summary>diff</summary>

```diff
diff --git a/src/box/box.cc b/src/box/box.cc
index 66f8c49a3..1f0528a70 100644
--- a/src/box/box.cc
+++ b/src/box/box.cc
@@ -90,7 +90,6 @@
 #include "flightrec.h"
 #include "wal_ext.h"
 #include "mp_util.h"
-#include "box_trigger_def.h"
 
 static char status[64] = "unconfigured";
 
@@ -416,7 +415,7 @@ box_do_set_orphan(bool orphan)
 	is_orphan = orphan;
 	box_update_ro_summary();
 	if (!is_orphan)
-		box_run_on_recovery_state("synced");
+		box_run_on_recovery_state(RECOVERY_STATE_SYNCED);
 }
 
 void
@@ -3772,7 +3771,7 @@ bootstrap_master(const struct tt_uuid *replicaset_uuid)
 	if (gc_checkpoint() != 0)
 		panic("failed to create a checkpoint");
 
-	box_run_on_recovery_state("snapshot_recovered");
+	box_run_on_recovery_state(RECOVERY_STATE_SNAPSHOT_RECOVERED);
 }
 
 /**
@@ -3822,7 +3821,7 @@ bootstrap_from_master(struct replica *master)
 					APPLIER_FINAL_JOIN;
 	applier_resume_to_state(applier, wait_state, TIMEOUT_INFINITY);
 
-	box_run_on_recovery_state("snapshot_recovered");
+	box_run_on_recovery_state(RECOVERY_STATE_SNAPSHOT_RECOVERED);
 
 	/*
 	 * Process final data (WALs).
@@ -4054,7 +4053,7 @@ local_recovery(const struct tt_uuid *instance_uuid,
 	 */
 	memtx_engine_recover_snapshot_xc(memtx, checkpoint_vclock);
 
-	box_run_on_recovery_state("snapshot_recovered");
+	box_run_on_recovery_state(RECOVERY_STATE_SNAPSHOT_RECOVERED);
 
 	engine_begin_final_recovery_xc();
 	recover_remaining_wals(recovery, &wal_stream.base, NULL, false);
@@ -4358,7 +4357,7 @@ box_cfg_xc(void)
 
 	rmean_cleanup(rmean_box);
 
-	box_run_on_recovery_state("wal_recovered");
+	box_run_on_recovery_state(RECOVERY_STATE_WAL_RECOVERED);
 
 	/* Follow replica */
 	replicaset_follow();
@@ -4399,8 +4398,8 @@ box_cfg_xc(void)
 	say_info("ready to accept requests");
 
 	if (is_bootstrap_leader) {
-		box_run_on_recovery_state("indexes_built");
-		box_run_on_recovery_state("synced");
+		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
+		box_run_on_recovery_state(RECOVERY_STATE_SYNCED);
 	}
 	if (!is_bootstrap_leader) {
 		replicaset_sync();
@@ -4455,6 +4454,42 @@ box_cfg(void)
 	}
 }
 
+/**
+ * Whether the recovery state triggers for a specific state have already run.
+ */
+static bool recovery_state_reached[RECOVERY_STATE_MAX];
+
+static const char *recovery_state_names[RECOVERY_STATE_MAX] = {
+	[RECOVERY_STATE_SNAPSHOT_RECOVERED] = "snapshot_recovered",
+	[RECOVERY_STATE_WAL_RECOVERED] = "wal_recovered",
+	[RECOVERY_STATE_INDEXES_BUILT] = "indexes_built",
+	[RECOVERY_STATE_SYNCED] = "synced",
+};
+
+int
+box_run_on_recovery_state(enum box_recovery_state state)
+{
+	/* Don't run triggers for the same stage twice. */
+	if (recovery_state_reached[state])
+		return 0;
+	recovery_state_reached[state] = true;
+	/*
+	 * on_recovery_state triggers expect a string, but passing a string
+	 * literal to them turns out quite ugly, because trigger_run takes
+	 * void *. So one ends up with
+	 * trigger_run(trig, (char *)"string_literal");
+	 *
+	 * Copying the string turns out less ugly and more safe, because the
+	 * resulting string becomes writable.
+	 */
+	assert(state >= 0 && state < RECOVERY_STATE_MAX);
+	char buf[64];
+	size_t len = strlcpy(buf, recovery_state_names[state], sizeof(buf));
+	assert(len < sizeof(buf));
+	(void)len;
+	return trigger_run(&box_on_recovery_state, buf);
+}
+
 /**
  * box.coredump() forks to save a core. The entire
  * server forks in box.cfg{} if background=true.
diff --git a/src/box/box.h b/src/box/box.h
index acb76f425..4a7afd57a 100644
--- a/src/box/box.h
+++ b/src/box/box.h
@@ -70,6 +70,47 @@ extern double on_shutdown_trigger_timeout;
 /** Invoked on box shutdown. */
 extern struct rlist box_on_shutdown_trigger_list;
 
+/**
+ * Triggers invoked during initial `box.cfg` call on various recovery stages.
+ */
+extern struct rlist box_on_recovery_state;
+
+/**
+ * Recovery states supported by on_recovery_state triggers.
+ * Are positioned in the order of appearance during initial box.cfg().
+ * The only exception are WAL_RECOVERED and INDEXES_BUILT, which might come in
+ * any order, since the moment when secondary indexes are built depends on
+ * box.cfg.force_recovery and box.cfg.memtx_use_mvcc_engine.
+ */
+enum box_recovery_state {
+	/**
+	 * The node has either recovered the snapshot from the disk, received
+	 * the snapshot from the remote master as part of initial join stage or
+	 * has bootstrapped the cluster.
+	 */
+	RECOVERY_STATE_SNAPSHOT_RECOVERED,
+	/**
+	 * The node has either recovered the local WAL files, received the WALs
+	 * from the remote master as part of final join or has bootstrapped the
+	 * cluster.
+	 */
+	RECOVERY_STATE_WAL_RECOVERED,
+	/**
+	 * The node has built secondary indexes for memtx spaces.
+	 */
+	RECOVERY_STATE_INDEXES_BUILT,
+	/**
+	 * The node has synced with remote peers. IOW, it has transitioned from
+	 * "orphan" to "running".
+	 */
+	RECOVERY_STATE_SYNCED,
+	RECOVERY_STATE_MAX,
+};
+
+/** Run the on_recovery_state_triggers with state \a state. */
+int
+box_run_on_recovery_state(enum box_recovery_state state);
+
 /**
  * Timeout during which the transaction must complete,
  * otherwise it will be rolled back.
diff --git a/src/box/box_trigger_def.h b/src/box/box_trigger_def.h
deleted file mode 100644
index 5e0fa438e..000000000
--- a/src/box/box_trigger_def.h
+++ /dev/null
@@ -1,36 +0,0 @@
-#pragma once
-/*
- * SPDX-License-Identifier: BSD-2-Clause
- *
- * Copyright 2022, Tarantool AUTHORS, please see AUTHORS file.
- */
-
-#include "trigger.h"
-
-/**
- * Triggers invoked during initial `box.cfg` call on various recovery stages.
- */
-extern struct rlist box_on_recovery_state;
-
-/** Run box_on_recovery_state triggers with state represented by \a str. */
-static inline int
-box_run_on_recovery_state(const char *str)
-{
-	/*
-	 * on_recovery_state triggers expect a string, but passing a string
-	 * literal to them turns out quite ugly, because trigger_run takes
-	 * void *. So one ends up with
-	 * trigger_run(trig, (char *)"string_literal");
-	 *
-	 * Compound literals solve this problem in a less ugly way, besides they
-	 * can really be written to:
-	 * trigger_run(trig, (char[]){"string_literal"}).
-	 * But unfortunately they aren't supported in c++.
-	 * So we are left with this:
-	 */
-	char buf[64];
-	size_t len = strlcpy(buf, str, sizeof(buf));
-	assert(len < sizeof(buf));
-	(void)len;
-	return trigger_run(&box_on_recovery_state, buf);
-}
diff --git a/src/box/lua/ctl.c b/src/box/lua/ctl.c
index 8d47fe69d..40d5ae4f2 100644
--- a/src/box/lua/ctl.c
+++ b/src/box/lua/ctl.c
@@ -40,7 +40,6 @@
 #include "lua/trigger.h"
 
 #include "box/box.h"
-#include "box/box_trigger_def.h"
 #include "box/schema.h"
 #include "box/engine.h"
 #include "box/memtx_engine.h"
diff --git a/src/box/memtx_engine.cc b/src/box/memtx_engine.cc
index 4049e3848..6a1e05aa2 100644
--- a/src/box/memtx_engine.cc
+++ b/src/box/memtx_engine.cc
@@ -56,7 +56,7 @@
 #include "read_view.h"
 #include "memtx_tuple_compression.h"
 #include "memtx_space.h"
-#include "box_trigger_def.h"
+#include "box.h"
 
 #include <type_traits>
 
@@ -437,7 +437,7 @@ memtx_engine_begin_final_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state("indexes_built");
+		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
 	}
 	return 0;
 }
@@ -456,7 +456,7 @@ memtx_engine_begin_hot_standby(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state("indexes_built");
+		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
 	}
 	return 0;
 }
@@ -476,7 +476,7 @@ memtx_engine_end_recovery(struct engine *engine)
 		memtx->state = MEMTX_OK;
 		if (space_foreach(memtx_build_secondary_keys, memtx) != 0)
 			return -1;
-		box_run_on_recovery_state("indexes_built");
+		box_run_on_recovery_state(RECOVERY_STATE_INDEXES_BUILT);
 	}
 	xdir_collect_inprogress(&memtx->snap_dir);
 
diff --git a/test/replication-luatest/recovery_triggers_test.lua b/test/replication-luatest/recovery_triggers_test.lua
index 943146f15..870990032 100644
--- a/test/replication-luatest/recovery_triggers_test.lua
+++ b/test/replication-luatest/recovery_triggers_test.lua
@@ -32,8 +32,7 @@ end)
 
 local function is_state_reached(server, state)
     return server:exec(function(state)
-        -- luacheck: ignore
-        for _, s in pairs(test_states) do
+        for _, s in pairs(_G.test_states) do
             if s == state then
                 return true
             end
@@ -95,4 +94,14 @@ g.test_recovery_stages = function(cg)
         }
     end)
     wait_state(cg.replica, 'synced')
+    -- Check that triggers on 'synced' stage are executed only once.
+    cg.replica:exec(function()
+        local t = require('luatest')
+        local num_events = #_G.test_states
+        local old_repl = box.cfg.replication
+        box.cfg{replication = ''}
+        box.cfg{replication = old_repl}
+        t.assert_equals(#_G.test_states, num_events,
+                        "Triggers on sync aren't re-run")
+    end)
 end
```

</details>